### PR TITLE
Variable width alignment

### PR DESCRIPTION
**Pull request by minad**
Not pixels but floating point values in units of char width are used here.
This is a bit simpler and probably flexible enough for our purposes.

### CODE REVIEW

**Comment by minad** [comment]
Test function, which was not possible before:

~~~ elisp
;; 1 char icon, 0.15 margin
(setq corfu-kind-formatter (lambda (_)
                             (propertize
                              (concat
                               (propertize " " 'display '(space :width 0.15))
                               "#"
                               (propertize " " 'display '(space :width 0.15)))
                              'corfu-width 1.3
                              'face '(:background "cyan"))))
~~~

**Comment by minad** [comment]
Alternatively one could also search for `(space :width float)` display properties but then the situation would get quickly out of hand, since the `:width` specifications can be so complicated (different units, formulas, etc).

File: corfu.el [hunk 1]
```diff
@@ -611,15 +611,20 @@ A scroll bar is displayed from LO to LO+BAR."
...
   (setq cands
         (cl-loop for c in cands collect
                  (cl-loop for s in c collect
                           (replace-regexp-in-string "[ \t]*\n[ \t]*" " " s))))
-  (let* ((cw (cl-loop for x in cands maximize (string-width (car x))))
-         (pw (cl-loop for x in cands maximize (string-width (cadr x))))
-         (sw (cl-loop for x in cands maximize (string-width (caddr x))))
+  (let* ((cw (cl-loop for x in cands maximize (corfu--string-width (car x))))
+         (pw (cl-loop for x in cands maximize (corfu--string-width (cadr x))))
+         (sw (cl-loop for x in cands maximize (corfu--string-width (caddr x))))
```
Comment (minad) [inline corfu.el:627]: I am overly general here - probably the 'corfu-width property should only be used for the prefix. The other strings won't have it.

**Comment by jdtsmith** [comment]
Time traveler and mind reader too, eh?  Quite the skillset ;).  I'll take a look.  An advantage of "I'm always 3 chars wide" is good centering for 1 char string, 2 char string or icon.

**Comment by minad** [comment]
> Time traveler and mind reader too, eh?

Just pretending to be a mind reader, easy as a time traveler :-P

> "I'm always 3 chars wide" is good centering for 1 char string, 2 char string or icon.

But seriously, this was just your idea since you wanted margin+icon+margin and the margin was reasonably close with 0.6. The only sacrifice we make is for 1 char text "icons" - a sacrifice which is justified in my opinion, since icons will be more popular.

So I close this variable width PR for now - keeping it simple, yay! We can still revisit this later if needed!

**Pull request closed without merging**